### PR TITLE
OpenPaaS-Suite/esn#16 - Add Dockerfile + Inject environment variable at runtime

### LINKED ISSUE #16: Update docker-compose.yml to deploy with the new frontends
Currently, our [docker-compose.yml](https://github.com/linagora/openpaas-esn/blob/master/docker-compose.yml) deployment system is not using the new frontends. This issue objective is to have a deployment of OpenPaaS through this docker-compose.yml file, so that it also deploy/uses the new frontends.

Related issue: https://github.com/OpenPaaS-Suite/esn/issues/28

### PR DESCRIPTION
**Pull request by piavgh**
Partially resolves OpenPaaS-Suite/esn#16.

## TODO:

- [x] Inject OPENPAAS_API_URL for login route

### CODE REVIEW

File: src/views/Login.vue [hunk 1]
```diff
@@ -47,8 +47,11 @@ export default {
   }),
   methods: {
     login() {
+      const baseUrl =
+        (window.openpaas && window.openpaas.OPENPAAS_API_URL) || "";
```
Comment (huy-ta) [inline src/views/Login.vue:51]:
```suggestion
      const baseUrl = (window.openpaas && window.openpaas.OPENPAAS_API_URL) || "";
```

Comment (itswillta) [inline src/views/Login.vue:57]:
```suggestion
      const baseUrl = (window.openpaas && window.openpaas.OPENPAAS_API_URL) || "";
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+FROM node:lts-alpine as build-stage
+
+WORKDIR /app
+
+COPY package*.json ./
+
+RUN npm install
+
+COPY . .
```
Comment (tuanlc) [inline Dockerfile:11]: Why do not have this at line 7? then you only need one `COPY`

File: scripts/generate_env.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+CONFIG_FILE="env/openpaas.js"
+
+mkdir -p env
+
+# Recreate config file
+rm -rf ./$CONFIG_FILE
```
Comment (tuanlc) [inline scripts/generate_env.sh]: no need `-r`, it is used to remove directory

File: scripts/generate_env.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+echo "window.openpaas = {" >> ./$CONFIG_FILE
+
+for key in $(env | awk -F "=" '{print $1}' | grep "OPENPAAS.*")
+# key is something like OPENPAAS_API_URL
+do
+    # Append configuration property to JS file
+    echo "  $key: \"$(printenv $key)\"," >> ./$CONFIG_FILE
+done
+
+echo "}" >> ./$CONFIG_FILE
```
Comment (tuanlc) [inline scripts/generate_env.sh:21]: Why not having a sample file and use `sed` to replace values?

File: Dockerfile (hunk 2 above)
Comment (piavgh) [reply Dockerfile:11]: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache

Please check the above URL, layer caching is one of the best practices in Docker image.
If we don't separate the 2 COPY(s), whenever we push a new change, Docker builds everything from that point without using cache.

If we separate them, only when we update `package.json` file (which we rarely do), the image is rebuilt from that point.

So it's used to reduce the build time.

File: scripts/generate_env.sh (hunk 4 above)
Comment (piavgh) [reply scripts/generate_env.sh:21]: Then we have to add that sample file in our git repository, which is redundant.
Creating a new file each time a container is started is much cleaner

File: src/views/Login.vue (hunk 1 above)
Comment (piavgh) [reply src/views/Login.vue:57]: Screenshot from 2020-08-11 09-17-39

Doing that and the ESLint Prettier of this project will scream :D. So I respected that current ESLint config of this project

File: scripts/generate_env.sh (hunk 4 above)
Comment (piavgh) [reply scripts/generate_env.sh:21]: Moreover, if we have a sample file (`env/openpaas.js`), we also need `copy-webpack-plugin` to copy this file (after `sed` with correct values) to the build.
Christophe also said no to this https://chat.linagora.com/openpaas/pl/14h5myek47r7zdu86j7syaws8h :D

File: src/views/Login.vue (hunk 1 above)
Comment (itswillta) [reply src/views/Login.vue:57]: The `printWidth` rule is probably too small. Well it doesn't really matter anyway.

File: Dockerfile (hunk 2 above)
Comment (tuanlc) [reply Dockerfile:11]: Good to know!

File: scripts/generate_env.sh (hunk 4 above)
Comment (tuanlc) [reply scripts/generate_env.sh:21]: Ok

**Review by itswillta — APPROVED** [review]

**Review by tuanlc — APPROVED** [review]

**Pull request merged**